### ISSUE TITLE
Retain cycle in HTMLNodeIterator

### ISSUE THREAD

**Issue by dsanghan** [iabudiab/HTMLKit#4]
When you do a simple doc.body, the document gets retained by the iterator in `root` and `referenceNode` and the iterator gets retained by the document in `attachNodeIterator` - hence neither gets deallocated.

```
- (instancetype)initWithNode:(HTMLNode *)node
				 showOptions:(HTMLNodeFilterShowOptions)showOptions
					  filter:(id<HTMLNodeFilter>)filter
{
	self = [super init];
	if (self) {
		_root = node;     // retains doc
		_filter = filter;     // retains doc
		_whatToShow = showOptions;
		_referenceNode = _root;
		_pointerBeforeReferenceNode	= YES;
		[_root.ownerDocument attachNodeIterator:self]; // doc retains iterator
	}
	return self;
}
```

A possible fix is to convert the `nodeIterator` array into a weak objects `NSHashTable`.

**Comment by iabudiab** [comment]
@dsanghan Hey, good catch 👍 Thanks for reporting.

**Comment by iabudiab** [comment]
The same bug affects `HTMLRange`. I'll open a new issue for those.

**Comment by iabudiab** [comment]
Fixed in 2.0.2
